rts/Game/Camera/SpringController.cpp

```diff
@@ -54,13 +63,41 @@ void CSpringController::ConfigUpdate()
 	fastScaleMousewheel = configHandler->GetFloat("CamSpringFastScaleMousewheelMove");
 	doRotate = configHandler->GetBool("CamSpringEdgeRotate");
 	lockCardinalDirections = configHandler->GetBool("CamSpringLockCardinalDirections");
+	trackMapHeight = configHandler->GetInt("CamSpringTrackMapHeightMode");
+
+	if (trackMapHeight == HeightTracking::Smooth && !modInfo.enableSmoothMesh) {
+		LOG_L(L_ERROR, "Smooth mesh disabled");
+		trackMapHeight = HeightTracking::Terrain;
+	}
 }
 
 void CSpringController::ConfigNotify(const std::string & key, const std::string & value)
 {
 	ConfigUpdate();
 }
 
+void CSpringController::SmoothCamHeight(const float3& prevPos) {
+	if (!pos.IsInBounds()) {
+		return;
+	}
+
+	float3 camPos = GetPos(); // new camera pos with height from previous frame
+	camPos.y = std::max(camPos.y, CGround::GetHeightReal(camPos.x, camPos.z, false) + 5.0f);
+
+	// raycast to ground to simulate camera movement and find new point of focus
+	const float distToGround = CGround::LineGroundCol(camPos, camPos + dir * 150000.0f, false);
+	// FIXME camera focus is now first ground intersection which is not what we want
```

Comment:
Also @dzosz perhaps you can consider using `readMap->GetUnsyncedHeightInfo(x_hm, z_hm)`, which stores min,max and average of a height-map patch 1024x1024 size in world coordinates. When the ray tracing is complete and you have a world coordinate where the ray hits the ground you can average the average or max height of some number of nearby "patches" and use that as a height guidance. Just an idea.

Reply:
Would it solve the following problem? Rotate the camera so the previous point of focus is hidden behind it. We don't know what exact (x,z) we want without raytracing through the hill. Raytracing multiple times would work but not guarantee that he last sucessful raytrace is the actual position we want (if the camera angle is very flat or map points downhill we might focus on the point.
![image](https://user-images.githubusercontent.com/13596691/232152155-f1037910-cc83-4c9f-9fcb-0f23718e2dfd.png)
where p1 is first point of focus, and p2 second point of focus (it should still be p1 though)